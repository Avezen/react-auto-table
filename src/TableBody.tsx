import React from 'react';
import TruncatedString from './TruncatedString';
// @ts-ignore
import loader from './assets/images/loader.svg'

interface TableBodyProps {
    data?: any[];
    isFetching: boolean;
    error?: boolean;
    colWidths?: number[];
    tableHead: string[];
    currentlyOpenedTd: {
        id: null,
        key: null,
    };
    toggleTd: any;
}

export const TableBody = ({
                          data,
                          isFetching,
                          error,
                          colWidths,
                          tableHead,
                          currentlyOpenedTd,
                          toggleTd,
                      }: TableBodyProps) => (
    <tbody>
    {isFetching ? (
        <tr>
            <td colSpan={100}>
                <div className={'fetch-info'}>
                    <img src={loader} alt={'loader'} width={30}/>
                </div>
            </td>
        </tr>
    ) : data && data.length > 0 ? (
        data.map((item: any, key: number) => {
            // let dateString = new Date(item.Deadline).toLocaleString('pl-PL', { hour12: false });
            // dateString = dateString.substr(0, dateString.indexOf(','));

            return (
                <tr key={key}>
                    {Object.entries(item).map(([objectKey, value]: [string, any]) => {
                        return (
                            <td key={objectKey}>
                                <TruncatedString
                                    toggled={item.Id === currentlyOpenedTd.id && key === currentlyOpenedTd.key}
                                    toggleTd={() => toggleTd(item.Id, key)}
                                    value={value}
                                />
                            </td>
                        );
                    })}

                    <td>
                        <button>Edit</button>
                    </td>
                </tr>
            );
        })
    ) : error ? (
        <tr>
            <td colSpan={100}>
                <div className={'fetch-info'}>Error while fetching data</div>
            </td>
        </tr>
    ) : (
        <tr>
            <td colSpan={100}>
                <div className={'fetch-info'}>Not found</div>
            </td>
        </tr>
    )}
    </tbody>
);
