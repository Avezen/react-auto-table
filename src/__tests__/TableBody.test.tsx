import React from "react";
import {TableBody, TableBodyProps} from "../TableBody";
import renderer from 'react-test-renderer';

function renderTableBody(props: Partial<TableBodyProps> = {}) {
    const defaultProps: TableBodyProps = {
        data: [
            {
                Id: 1,
                Name: 'Test 1'
            },
            {
                Id: 2,
                Name: 'Test 2'
            }
        ],
        isFetching: false,
        error: false,
        colWidths: [50, 100],
        tableHead: ['Id', 'Name'],
        currentlyOpenedTd: {
            id: 1,
            key: 'Name',
        },
        toggleTd() {
            return;
        }
    };

    return <TableBody {...defaultProps} {...props} />;
}

describe("<TableBody />", () => {
    test("should render component with data", async () => {
        const findByTestId = renderTableBody(),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });

    test("should render component with data, with toggled tooltip", async () => {
        const props = {
            data: [
                {
                    Id: 1,
                    Name: 'Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, '
                },
                {
                    Id: 2,
                    Name: 'Test 2'
                }
            ],
        };
        const findByTestId = renderTableBody(props),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });

    test("should render component with data", async () => {
        const findByTestId = renderTableBody(),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });

    test("should render component with error", async () => {
        const props = {
            data: [],
            error: true
        };
        const findByTestId = renderTableBody(props),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });

    test("should render component while fetching", async () => {
        const props = {
            data: [],
            isFetching: true,
            error: true
        };
        const findByTestId = renderTableBody(props),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });

    test("should render component with no data", async () => {
        const props = {
            data: []
        };
        const findByTestId = renderTableBody(props),
            TableBodyComponent = renderer.create(findByTestId).toJSON();

        expect(TableBodyComponent).toMatchSnapshot();
    });
});
