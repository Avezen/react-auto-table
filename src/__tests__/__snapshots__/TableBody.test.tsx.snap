// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableBody /> should render component while fetching 1`] = `
<tbody>
  <tr>
    <td
      colSpan={100}
    >
      <div
        className="fetch-info"
      >
        <img
          alt="loader"
          src="test-file-stub"
          width={30}
        />
      </div>
    </td>
  </tr>
</tbody>
`;

exports[`<TableBody /> should render component with data 1`] = `
<tbody>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          1
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 1
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          2
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 2
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
</tbody>
`;

exports[`<TableBody /> should render component with data 2`] = `
<tbody>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          1
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 1
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          2
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 2
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
</tbody>
`;

exports[`<TableBody /> should render component with data, with toggled tooltip 1`] = `
<tbody>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          1
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Te...
        </div>
        <div
          className="custom-tooltip"
          data-testid="expand-tooltip"
          onClick={[Function]}
        >
          <img
            alt="expand"
            data-testid="expand-icon"
            src="test-file-stub"
            width={15}
          />
          <span
            className="tooltip-text visible"
            data-testid="full-value"
          >
            Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, Test 1, 
          </span>
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
  <tr>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          2
        </div>
      </div>
    </td>
    <td>
      <div
        className="truncated-string"
      >
        <div
          data-testid="truncated-value"
        >
          Test 2
        </div>
      </div>
    </td>
    <td>
      <button>
        Edit
      </button>
    </td>
  </tr>
</tbody>
`;

exports[`<TableBody /> should render component with error 1`] = `
<tbody>
  <tr>
    <td
      colSpan={100}
    >
      <div
        className="fetch-info"
      >
        Error while fetching data
      </div>
    </td>
  </tr>
</tbody>
`;

exports[`<TableBody /> should render component with no data 1`] = `
<tbody>
  <tr>
    <td
      colSpan={100}
    >
      <div
        className="fetch-info"
      >
        Not found
      </div>
    </td>
  </tr>
</tbody>
`;
